adal/src/main/java/com/microsoft/aad/adal/BrokerProxy.java

```diff
@@ -116,19 +116,25 @@ public BrokerProxy(final Context ctx) {
     public boolean canSwitchToBroker() {
         String packageName = mContext.getPackageName();
 
-        // ADAL switches broker for following conditions:
-        // 1- app is not skipping the broker
-        // 2- permissions are set in the manifest,
-        // 3- if package is not broker itself for both company portal and azure
-        // authenticator
-        // 4- signature of the broker is valid
-        // 5- account exists
-        return AuthenticationSettings.INSTANCE.getUseBroker() 
-                && verifyManifestPermissions()
-                && checkAccount(mAcctManager, "", "")
-                && !packageName.equalsIgnoreCase(AuthenticationSettings.INSTANCE.getBrokerPackageName())
-                && !packageName.equalsIgnoreCase(AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME)
-                && verifyAuthenticator(mAcctManager);
+        try {
+            // ADAL switches broker for following conditions:
+            // 1- app is not skipping the broker
+            // 2- permissions are set in the manifest,
+            // 3- if package is not broker itself for both company portal and azure
+            // authenticator
+            // 4- signature of the broker is valid
+            // 5- account exists
+            return AuthenticationSettings.INSTANCE.getUseBroker()
+                    && verifyManifestPermissions()
```

Comment:
1. rename to  "CheckForPermissionAPI23orAbove", if API > 22 && other four checks get passed && permissionNotGranted, then return error. Here by return error, throw UsageAuthenticationException and pass it to callback
2. this verification should be the last one to check